### ISSUE TITLE
Addition of L1 norm for sparse weights

### ISSUE THREAD

**Issue by razor1179** [mravanelli/pytorch-kaldi#67]
I am currently trying to implement group lasso and l1 norm regularization to compare different sparsity methods. I am trying to accomplish this by adding an operation named `cost_l1` in `utils.py`. I am not sure how to proceed. Any tips would be helpful.

**Comment by mravanelli** [comment]
Hi!,
in PyTorch the weight decay normalization is defined in the optimizer
directly (see opt_weight_decay parameters in the config file).
Unfortunately, only l2 normalization is natively implemented. If you would
like to implement a different regularization scheme (e.g. lasso or l1) I
would suggest you to go into the *core.py* function and manually add your
regularization penalty before doing back-propagation. To do it, take a look
into the following line:

outs_dict['loss_final'].backward()

You can change it in this way:
outs_dict['loss_final']=outs_dict['loss_final']+ 0.01 * reg_penalty.

To compute the reg_penalty you probably need access to the weights of the
neural network.
You can find all the neural parameters in the dictionary called "nns".

Please, let me know if this suggestion could be useful for your application.

Best,

Mirco

**Comment by razor1179** [comment]
@mravanelli thank you very much for the suggestion, I did however figure another way out, in the `[model]` section in the *.cfg I included either of the two loss functions `loss_gl` or `loss_l1`. An example below is how I implemented it for group Lasso:

```
[model]
model_proto = proto/model.proto
model = out_dnn1=compute(LSTM_cudnn_layers,fmllr)
	out_dnn2=compute(MLP_layers,out_dnn1)
	out_dnn3=compute(MLP_layers2,out_dnn1)
	loss_mono=cost_nll(out_dnn3,lab_mono)
	loss_mono_w=mult_constant(loss_mono,1.0)
	loss_cd=cost_nll(out_dnn2,lab_cd)
	loss_gl=cost_gl(out_dnn2,lab_cd)
	loss_cd_l1=sum(loss_gl,loss_cd)
	loss_final=sum(loss_cd_l1,loss_mono_w)
	err_final=cost_err(out_dnn2,lab_cd)
```
I used the sum operation to accumulate the group Lasso loss into the final_loss, in the `model.proto` I added the two losses:
```
cost_l1(input,label)
cost_gl(input,label)
```
The `input` and `label` options are present as that is the format expected and I couldn't find a way around it.

In `utils.py` I added two functions as below:
```
def l1_norm(model):
    l1_reg = torch.tensor(0, dtype=torch.float32).cuda()
    for key in model:
        for param in model[key].parameters():
            dim = param.size()
            if dim.__len__() > 1:
                l1_reg += torch.norm(param, 1)
    return l1_reg * 0.008

def gl_norm(model, num_blk):
    gl_reg = torch.tensor(0., dtype=torch.float32).cuda()
    for key in model:
        for param in model[key].parameters():
            dim = param.size()
            if dim.__len__() > 1:
                div1 = list(torch.chunk(param,num_blk,1))
                all_blks = []
                for div2 in div1:
                    temp = list(torch.chunk(div2,num_blk,0))
                    for blk in temp:
                        all_blks.append(blk)
                for l2_param in all_blks:
                    gl_reg += torch.norm(l2_param, 2)
    return gl_reg * 0.009
```
Also I made changes in `model_init` by adding:
```
        if operation == 'cost_l1':
            inp_out_dict[out_name] = [1]

        if operation == 'cost_gl':
            inp_out_dict[out_name] = [1]
```
and in `forward_model` by adding:
```
       if operation == 'cost_l1':
            if to_do != 'forward':
                outs_dict[out_name] = l1_norm(nns)

        if operation == 'cost_gl':
            if to_do != 'forward':
                outs_dict[out_name] = gl_norm(nns, 16)
```

These changes enables the implementation of L1 regularization as well as group Lasso but the lambda value currently needs to be changed manually in the `l1_norm` or `gl_norm` function.  @TParcollet, please let me know if you can see a more efficient way to implement the two functions.
